Virtualization default empty states (#466)
openshift/openshift-origin-design: designs/virtualization/4.6/empty-states/empty-states.md

# Virtualization Default Empty States
Foday Kargbo - UXD
## Background

> Virtualization is being introduced to the console. Virtualization will be accessed via the main left navigation panel. Once the user navigates to Virtualization, they will see 2 tabs, one for the Virtual Machines screen, and one for the Virtual Machine Templates screen. The Virtual Machines tab will be selected by default.

Reviewer: ```suggestion
Virtualization is installed via an operator in the console. Once installed,  Virtualization will be accessed via the main left navigation panel. Once the user navigates to Virtualization, they will see 2 tabs, one for the Virtual Machines screen, and one for the Virtual Machine Templates screen. The Virtual Machines tab will be selected by default.
```

## Objectives
Users who have not created virtual machines or virtual machine templates will not be able to see any data on these screens. 

> TBD: Do we want to create an empty state screen for the virtualization dashboard?

Author: Sounds good.

> TBD: Do we want to create an empty state screen for the virtualization dashboard?

Author: Sounds good!

The goals of this task:
- Create empty state screens when there is no data to show
- Inform the user about the current state
- Educate the user about what they will see when data is available
- Guide the user to take action
- Explore any useful content that might help the user be contextually productive?
## 1. Virtual machine default empty state screen

> ![Virtual machines empty state screen](img/virtual-machines-empty-state.png "Virtual machines empty state screen")

Author: I'll investigate this and fix it.

> ![Virtual machines empty state screen](img/virtual-machines-empty-state.png "Virtual machines empty state screen")

Author: Fixed images. You can new view them.

> ![Virtual machines empty state screen](img/virtual-machines-empty-state.png "Virtual machines empty state screen")

Author: Great! I like the wording, we'll go with that. I agree that it should reflect or draw attention to the filtered view. I'm curious to know which case would this be? In other words, how is the filter determined for someone who has not created any VMs? We can adjust the wording based on this logic or we could go with a standard copy as your first suggestion. Thanks!

> ![Virtual machines empty state screen](img/virtual-machines-empty-state.png "Virtual machines empty state screen")

Reviewer: I wonder if this would be dynamic when they are on all projects. We'd want to ensure if we go this route that it can detect that and adjust accordingly.

> ![Virtual machines empty state screen](img/virtual-machines-empty-state.png "Virtual machines empty state screen")

Author: I'll investigate this with the developer team.

> ![Virtual machines empty state screen](img/virtual-machines-empty-state.png "Virtual machines empty state screen")

Author: @matthewcarleton I discussed this with Tomas and it is possible to detect which filter is on and can be implemented. We can change the text dynamically to reflect the active filter. I recommend we alternate the copy dynamically to add context to the user's experience.

Confirming that for "all projects" we will use this copy:

    No virtual machines
    There are no virtual machines available. Create a virtual machine to continue.

and when filtered to a specific project we will use this copy:

No virtual machines
There are no virtual machines available in this project. Create a virtual machine to continue.

I appreciate the input @matthewcarleton @andybraren

> ![Virtual machines empty state screen](img/virtual-machines-empty-state.png "Virtual machines empty state screen")

Reviewer 2: +1 @matthewcarleton I think it makes it clearer when you add 'across all projects' as oppose to not mentioning where.
And maybe we can just add the word 'Please' before the 'Create a virtual machine to continue'. IMO it makes it sound more pleasant :)

> ![Virtual machines empty state screen](img/virtual-machines-empty-state.png "Virtual machines empty state screen")

Author: @yfrimanm Yes, something we can consider. Thanks for the input!

## UI Inventory
1. Tabs
2. Graphic (Icon)
3. Title text
4. Body text
5. Call to action button
6. Secondary call to action (optional)

## Discussion

> 1. Tabs: Are we ok with the title on this tab? Yes (Resolved)

Reviewer: I think the discussion/open questions should be in the description of this PR. What we aim for is a merged design with no open questions.

Author: Ok, I'll move them to the description section.

2. Graphic: We should use an image with context 
3. Title text: Are we ok with this title?
4. Body text: Are we ok with this text?
5. Call to action button: On click/tap it opens a dropdown menu with 3 options:
    - Create via option 1
    - Create via option 2
    - Create via option 3

> 6. Secondary call to action: (optional) Is there a need for this? We could use this as an opportunity to educate the user. Alternatively, we could use this to present a secondary option when creating a virtual machine.

Reviewer: I think we really want/need to push quick starts here. We could even have a more prominent presentation of them if the user hasn't created anything yet. 
With this new push to provide images for an OS source when the operator ships we could either point users to create a custom VM OR create from a template. It could be worth exploring what it looks like on this page to have both.

Author: Ok, I'll explore some directions for how we could present quick starts to the use here.

Author: I've added two screens showing possible placements for Quick Starts and Create from Template.

## 2. Virtual machine templates default empty state screen

> ![Virtual machine templates empty state screen](img/virtual-machine-templates-empty-state.png "Virtual machine templates empty state screen")

Author: I like this as well. I'll research the filter issue and adjust accordingly. Thanks!

> ![Virtual machine templates empty state screen](img/virtual-machine-templates-empty-state.png "Virtual machine templates empty state screen")

Reviewer: A good question to raise to the developer team on our side (to start) and then to the general console team if they don't know :)

> ![Virtual machine templates empty state screen](img/virtual-machine-templates-empty-state.png "Virtual machine templates empty state screen")

Author: Will do.

> ![Virtual machine templates empty state screen](img/virtual-machine-templates-empty-state.png "Virtual machine templates empty state screen")

Reviewer 3: This is what GitHub is showing when it can't find what you're looking for
![image](https://user-images.githubusercontent.com/65043359/92488959-e0dcbc00-f1f7-11ea-869e-75feaf7abb2d.png)

I don't know if this is our brands' tone, but I like it! It's more humane and made me smile (:

## UI Inventory
1. Tabs
2. Graphic (Icon)
3. Title text
4. Body text
5. Call to action button
6. Secondary call to action (optional)

## Discussion
1. Tabs: Are we ok with the title on this tab?
2. Graphic: We should use an image with context 
3. Title text: Are we ok with this title?
4. Body text: Are we ok with this text?
5. Call to action button: On click/tap it opens a dropdown menu with the pertinent options
6. Secondary call to action: (optional) Is there a need for this? Suggestion: We could use this as an opportunity to educate the user. Alternatively, we could use this to present a secondary option when creating virtual machine templates.
## Observations

> **Note:** While researching this task, I noticed that we have empty state design conventions both in PatternFly and OpenShift Design. This empty state convention is not being utilized on the console currently. We should implement it as a default for all full screen empty state instances across the console for empty states.

Reviewer: I think we can pull this out from the PR.

Author: Ok, I'll remove this piece.

> # Virtualization Default Empty States

Reviewer: ```suggestion
---
parent: Virtualization
version: 4.7
---
# Virtualization Default Empty States
```

## 2. Virtual machine templates default empty state screen
![Virtual machine templates empty state screen](img/virtual-machine-templates-empty-state.png)
## UI Inventory
1. Tabs
2. Graphic (Icon)
3. Title text
4. Body text
5. Call to action button
6. Secondary call to action (optional)

## Discussion
1. Tabs: Are we ok with the title on this tab?
2. Graphic: We should use an image with context. Comments? 
3. Title text: Are we ok with this title?
4. Body text: Are we ok with this text?
5. Call to action button: On click/tap it opens a dropdown menu with 2 options:
    - New with Wizard
    - New from YAML
6. Secondary call to action: (optional) Is there a need for this? Suggestion: We could use this as an opportunity to educate the user. This could be a good place for a "Quick Start" button. Alternatively, we can introduce the Quick Start above the tabs. Quick Start placement is a work-in-progress and still needs to be confirmed. Comments?

## Explorations
Exploring various ways to push Quick Starts in the empty states.
### Screen 1

> ![Virtual machine empty state screen](img/virtual-machines-empty-state-v2.png)

Reviewer: This would be an action to create from a template right? Not to see a quick start about creating from a template? In any case we'll want to add the "New from template" option in the action menu - we don't currently show this "new" tag for things like this but it's not a bad idea. Something worth bringing up to the wider group.

Author: @matthewcarleton Yes, that is correct. We could add that action into the menu to focus all primary actions into one place (i.e. the CTA button).

Yes, I thought the 'New' tag could potentially help existing users note new features introduced on the platform.

### Screen 2

> ![Virtual machine empty state screen 2](img/virtual-machines-empty-state-v3.png)

Reviewer: for this action I think we'd want to ensure the user knows it's a quick start and not just a normal action. It could be something like "Learn how to create from template" or something like that. WDYT?

Author: Yes, some microcopy will be helpful to the user. I'm exploring some options now.

Author: I'm also curious about whether we're going to have a standardized 'Quick Start' component across the console for consistency...?

3. Title text: Are we ok with this title?
4. Body text: Are we ok with this text?
5. Call to action button: On click/tap it opens a dropdown menu with 3 options:
    - New with Wizard
    - New from YAML
    - New from Template
6. Secondary call to action: (optional) Is there a need for this? Suggestion: We could use this as an opportunity to educate the user. This could be a good place for a "Quick Start" button. Alternatively, we can introduce the Quick Start above the tabs. Quick Start placement is a work-in-progress and still needs to be confirmed. Comments?

## 2. Virtual machine templates default empty state screen
![Virtual machine templates empty state screen](img/virtual-machine-templates-empty-state.png)
## UI Inventory
1. Tabs
2. Graphic (Icon)
3. Title text
4. Body text
5. Call to action button
6. Secondary call to action (optional)

## Discussion
1. Tabs: Are we ok with the title on this tab?
2. Graphic: We should use an image with context. Comments? 
3. Title text: Are we ok with this title?
4. Body text: Are we ok with this text?
5. Call to action button: On click/tap it opens a dropdown menu with 2 options:
    - New with Wizard
    - New from YAML

> 6. Secondary call to action: (optional) Is there a need for this? Suggestion: We could use this as an opportunity to educate the user. This could be a good place for a "Quick Start" button. Alternatively, we can introduce the Quick Start above the tabs. Quick Start placement is a work-in-progress and still needs to be confirmed. Comments?

Reviewer: ```suggestion
6. Secondary call to action: For OpenShift Virtualization we will make use of the secondary action to call attention to the Quick starts that are relevant to them, this will take them to a filtered view of the virtualization quick starts.
```
What do you think? We'll need to confirm for 4.7 quick starts will have the ability to filter. I think just having a generic link off to quick starts is appropriate for the empty state because we won't know which quick start is most relevant to them.

Reviewer 2: makes sense to me as well, even though that maybe clicking a link might be less 'Call to action' than having the QS "in your face"

Reviewer: Based on what @gdoyle1 presented at the ecosystem this wouldn't filter the quick start page but show the side panel with just virtualization quick starts. We should adjust this content.

## Explorations
Exploring various ways to push Quick Starts in the empty states.
### Screen 1
![Virtual machine empty state screen 1](img/virtual-machines-empty-state-v2.png)

### Screen 2
![Virtual machine empty state screen 2](img/virtual-machines-empty-state-v3.png)

> ### Screen 2.1 

Reviewer: I think we could use this image to replace what you have above so we are proposing two separate designs. One thing we should also consider is how long these "new" badges last. I guess they'd be there for a whole release right?
